crucible/src/Lang/Crucible/Backend/ProofGoals.hs

```diff
@@ -55,16 +59,68 @@ newtype FrameIdentifier = FrameIdentifier Word64
  deriving(Eq,Ord,Show)
 
 
--- | A data-strucutre that can incrementally collect goals in context.
+-- | A data-structure that can incrementally collect goals in context.
 --   It keeps track both of the collection of assumptions that lead to
 --   the current state, as well as any proof obligations incurred along
 --   the way.
+--
+--   The main use of 'GoalCollector' is as the state of an
+--   'Lang.Crucible.Backend.AssumptionStack.AssumptionStack', which itself is
+--   part of the state of the simple and online backends.
+--
+--   'GoalCollector' can be somewhat counter-intuitive. The "top"
+--   ('TopCollector') is at the *leaves* when 'GoalCollector' is considered as a
+--   tree (which is a common way to conceptualize recursive algebraic data types
+--   such as this one). Furthermore, the frame identified by the frist argument
+--   of 'CollectorFrame' does not conceptually contain the goals *inside* the
+--   second ('GoalCollector') argument, but rather contains all the assumptions
+--   and goals in whatever 'GoalCollector' *contains* the 'CollectorFrame'
+--   constructor (everything *outside* of the 'CollectorFrame').
```

Comment:
I would be interested to know if (1) these facts are counter-intuitive to anyone besides me and (2) if this seems like a good way to clarify them.

Reply:
Rad!